File: hrms/src/main/java/com/roima/hrms/Service/Interfaces/NotificationService.java
package com.roima.hrms.Service.Interfaces;

import com.roima.hrms.Core.Entities.Notification;
import com.roima.hrms.Core.Entities.User;
import com.roima.hrms.Core.Enums.EntityType;
import com.roima.hrms.Core.Enums.NotificationType;
import com.roima.hrms.Dtos.notificationResponseDto;
import org.springframework.web.servlet.mvc.method.annotation.SseEmitter;

import java.util.List;
import java.util.UUID;

public interface NotificationService {

    void createNew(User recipient, User actor, NotificationType type, String title, String message);

    SseEmitter addEmitter(UUID userId);

    void notifyUser(UUID userId, notificationResponseDto response);

    void markAsRead(UUID notificationId);

    List<notificationResponseDto> getUnreadNotifications(UUID userId);

}

File: client/src/services/notificationService.ts
import { apiClient } from "./apiClient";
import type { ApiResponse } from "@/types/http";

export type NotificationResponseDto = {
  id: string;
  actor: string;
  recipient: string;
  type: string;
  title: string;
  message: string;
  created_at: string;
  read: boolean;
};

type ApiResult<T> = Promise<ApiResponse<T>>;

class NotificationService {
  markAsRead(notificationId: string): ApiResult<void> {
    return apiClient.put<void>(`/notifications/${notificationId}/read`);
  }

  getUnreadNotifications(): ApiResult<NotificationResponseDto[]> {
    return apiClient.get<NotificationResponseDto[]>(`/notifications/unread`);
  }
}

export const notificationService = new NotificationService();
